aiida/orm/nodes/data/jsonable.py

```diff
@@ -0,0 +1,144 @@
+            dictionary['@module'] = obj.__class__.__module__
+
+        # Even though the dictionary returned by ``as_dict`` should be JSON-serializable and therefore this should be
+        # sufficient to be able to generate a JSON representation and thus store it in the database, there is a
+        # difference in the JSON serializers used by Python's ``json`` module and those of the PostgreSQL database that
+        # is used for the database backend. Python's ``json`` module automatically serializes the ``inf`` and ``nan``
+        # float constants to the Javascript equivalent strings, however, PostgreSQL does not. If we were to pass the
+        # dictionary from ``as_dict`` straight to the attributes and it were to contain any of these floats, the storing
+        # of the node would fail, even though technically it is JSON-serializable using the default Python module. To
+        # work around this asymmetry, we perform a serialization round-trip with the ``JsonEncoder`` and ``JsonDecoder``
+        # where in the deserialization, the encoded float constants are not deserialized, but instead the string
+        # placeholders are kept. This now ensures that the full dictionary will be serializable by PostgreSQL.
+        try:
+            serialized = json.loads(json.dumps(dictionary), parse_constant=lambda x: x)
+        except TypeError as exc:
+            raise TypeError(f'the object `{obj}` is not JSON-serializable and therefore cannot be stored.') from exc
+
+        self.set_attribute_many(serialized)
+
+    @classmethod
+    def _deserialize_float_constants(cls, data: typing.Any):
+        """Deserialize the contents of a dictionary ``data`` deserializing infinity and NaN string constants.
+
+        The ``data`` dictionary is recursively checked for the ``Infinity``, ``-Infinity`` and ``NaN`` strings, which
+        are the Javascript string equivalents to the Python ``float('inf')``, ``-float('inf')`` and ``float('nan')``
+        float constants. If one of the strings is encountered, the Python float constant is returned and otherwise the
+        original value is returned.
+        """
+        if isinstance(data, dict):
+            return {k: cls._deserialize_float_constants(v) for k, v in data.items()}
+        if isinstance(data, list):
+            return [cls._deserialize_float_constants(v) for v in data]
+        if data == 'Infinity':
+            return float('inf')
+        if data == '-Infinity':
+            return -float('inf')
+        if data == 'NaN':
+            return float('nan')
+        return data
+
+    def _get_object(self) -> JsonSerializableProtocol:
+        """Return the cached wrapped object.
+
+        .. note:: If the object is not yet present in memory, for example if the node was loaded from the database,
+            the object will first be reconstructed from the state stored in the node attributes.
+
+        """
+        try:
+            return self._obj
```

Comment:
actually, wait when would this raise an `AttributeError`, do we not want to always reload the object, in case the attributes have changes?

Reply:
Ah yeh the attribute error would be for if its loaded I guess, but yeh what if you did:

```
data = JsonableData(obj)
data.set_attrbiute('x', 1)
data.obj
```

would the user expect the obj to be updated?

Reply:
Good point. This is only really a problem though for unstored nodes and I doubt that users will be manually changing the attributes of the serialized object. Since the whole point of this class is to allow simple wrapping of serializable instances, there seems little point to be manually changing the attributes after construction but before storing. If we do want to support this as a use case, then we should indeed always fully reconstruct the obj from the database. This will introduce a significant performance hit since it needs to hit the db and deserialize the entire object. I would argue that this is not worth it, since these nodes will used most often to retrieve the `obj` from a stored node, rather than mutating unstored versions of them.

TL;DR: I would keep the current implementation and at most just add a warning in the `obj` property docstring that the thing is cached and so messing with attributes manually can break things

Reply:
Yeh that sounds fine to me 👍